ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/MacOSX/Cocoa/procinfo.mm

```diff
@@ -0,0 +1,157 @@
+  NSString *wids = [[NSString alloc] init];
+  const CGWindowLevel kScreensaverWindowLevel = CGWindowLevelForKey(kCGScreenSaverWindowLevelKey);
+  CFArrayRef windowArray = CGWindowListCopyWindowInfo(kCGWindowListOptionAll, kCGNullWindowID);
+  CFIndex windowCount = 0;
+  if ((windowCount = CFArrayGetCount(windowArray))) {
+    for (CFIndex i = 0; i < windowCount; i++) {
+      NSDictionary *windowInfoDictionary =
+      (__bridge NSDictionary *)((CFDictionaryRef)CFArrayGetValueAtIndex(windowArray, i));
+      NSNumber *ownerPID = (NSNumber *)(windowInfoDictionary[(id)kCGWindowOwnerPID]);
+      NSNumber *level = (NSNumber *)(windowInfoDictionary[(id)kCGWindowLayer]);
+      if (level.integerValue < kScreensaverWindowLevel) {
+        NSNumber *windowID = windowInfoDictionary[(id)kCGWindowNumber];
+        if (pid == ownerPID.integerValue) {
+          wids = [wids stringByAppendingString:[@(windowID.integerValue) stringValue]];
+          wids = [wids stringByAppendingString:@"|"];
+        }
+      }
+    }
+  }
+  if ([wids length] > 0) {
+    wids = [wids substringWithRange:NSMakeRange(0, [wids length] - 1)];
+  }
+  const char *result = [wids UTF8String];
+  CFRelease(windowArray);
+  [wids release];
+  return result;
+}
+
+unsigned long cocoa_get_wid_or_pid(bool wid) {
+  unsigned long result;
+  const CGWindowLevel kScreensaverWindowLevel = CGWindowLevelForKey(kCGScreenSaverWindowLevelKey);
+  CFArrayRef windowArray = CGWindowListCopyWindowInfo(kCGWindowListOptionAll, kCGNullWindowID);
+  CFIndex windowCount = 0;
+  if ((windowCount = CFArrayGetCount(windowArray))) {
+    for (CFIndex i = 0; i < windowCount; i++) {
+      NSDictionary *windowInfoDictionary =
+      (__bridge NSDictionary *)((CFDictionaryRef)CFArrayGetValueAtIndex(windowArray, i));
+      NSNumber *ownerPID = (NSNumber *)(windowInfoDictionary[(id)kCGWindowOwnerPID]);
+      NSNumber *level = (NSNumber *)(windowInfoDictionary[(id)kCGWindowLayer]);
+      if (level.integerValue == 0) {
+        NSNumber *windowID = windowInfoDictionary[(id)kCGWindowNumber];
+        result = wid ? windowID.integerValue : ownerPID.integerValue;
+        break;
+      }
+    }
+  }
+  CFRelease(windowArray);
+  return result;
+}
+
+void cocoa_wid_to_top(CGWindowID wid) {
+  CFIndex appCount = [[[NSWorkspace sharedWorkspace] runningApplications] count];
+  for (CFIndex i = 0; i < appCount; i++) {
+    NSWorkspace *sharedWS = [NSWorkspace sharedWorkspace];
+    NSArray *runningApps = [sharedWS runningApplications];
+    NSRunningApplication *currentApp = [runningApps objectAtIndex:i];
+    pid_t currentPid = [currentApp processIdentifier];
+    if (cocoa_pid_from_wid(wid) == currentPid) {
+      NSRunningApplication *appWithPID = currentApp;
+      NSUInteger options = NSApplicationActivateAllWindows;
+      options |= NSApplicationActivateIgnoringOtherApps;
+      [appWithPID activateWithOptions:options];
+      break;
+    }
+  }
+}
+
+void cocoa_wid_set_pwid(CGWindowID wid, CGWindowID pwid) {

```

Comment:
@RobertBColton I'm especially proud of myself for figuring out a way to do this on macOS. Basically, it allows any window regardless of what process it belongs to become like a modal dialog in that it will always stay on top of a given parent window, provided the parent window belongs to the current app. This is useful for a number of reasons, and example is you could have a game that runs in windowed mode which has a main menu with a "more games" button. When that button is selected, you could open an internet browser with your games website or store as the default URL. Then you can get the browser's window id and from that information make it stay on top of your game until the browser is closed, so you won't have to Alt+Tab if you lose focus. (Mac uses a different key combination for task switching, but that's not important). Similarly can be done with opening a readme.txt in the default text editor. Believe it or not, there are a lot of practical uses for a lot of these functions, and I'm only explaining the usefulness of just one of them. This function was a nightmare to figure out on Mac, they make you jump through a lot of hoops to do things that involve working outside of the current process.